pkg/v1/abstractions/abstractions.go

```diff
@@ -0,0 +1,65 @@
+package abstractions
+
+import (
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+// Better names for these might be ArtifactSource or ArtifactSink.
+
+// Source represents an abstraction from which we can read artifacts.
+//
+// Descriptors() and Tags() are useful for discovering what's available from a Source.
+// Image(), Index(), and Layer() are useful for actually fetching content from that Source.
+//
+// For example:
+// * a tarball can contain multiple images, exposing its contents via Tags.
+// * a remote repository can act as a source, exposing its contents via Tags (/tags/list).
+// * any image index can act as a source, exposing its children via Descriptors.
+// * an OCI image layout acts as a source, exposing contents of index.json via Descriptors.
```

Comment:
Not sure how to articulate the difference, but I feel tarball/remote (storage) and image/index (artifact) are very different. Mixing them all together as `Source` appears confusing to me. And usually you know what storage you're dealing with while you don't know what artifact type you will get. And I think you're trying to solve the "don't know what artifact type" problem, not the other one.

Reply:
Certainly the `image index` example is just a curiosity. An image source is in practice very similar to an image index.

With the OCI image layout, I think it's more similar to remote/tarball because it's a storage format. It just so happens to be that the OCI image layout is stored as an Image Index + a directory of blobs.

They're very similar, so it *might* be useful to expose something that can turn an ImageIndex into a Source (or vice versa) just to satisfy the type system, but they're not quite the same.

Reply:
> usually you know what storage you're dealing with

this is the problem I'm trying to solve with `Sink`. We'd like to write generic algorithms that do stuff to images, where we don't care what the source is and we don't care what the sink is. We just want to e.g. copy them, append a ko binary to them, etc.

> don't know what artifact type

That's what the ImageDescriptor thing would solve, I think. It's just that they're kind of similar. There's definitely multiple problems being tackled here, so it's not super clear.

If we started with adding e.g. `remote.ImageDescriptor(name.Reference)`, we'd be able to handle this ambiguous artifact case. It's then pretty easy to assemble a remote `Source` from `remote.{Image,Index,ImageDescriptor}`

Reply:
> With the OCI image layout, I think it's more similar to remote/tarball because it's a storage format. It just so happens to be that the OCI image layout is stored as an Image Index + a directory of blobs.

Why OCI image layout is a storage format? It only describes the referential relation but doesn't say how manifests or blobs are actually fetched/stored. But tarball and remote are different, e.g. tarball goes to some file path to fetch blobs; remote goes to some `{registry}/{repo}/blobs/{ref}` to fetch blobs. That's why I think they should be at a lower layer than image/index are.

Reply:
https://github.com/opencontainers/image-spec/blob/master/image-layout.md

Reply:
Ah, I thought the image layout is the layout of image manifest... Apparently it's a real storage format. Is image layout being used anywhere nowadays?

Reply:
Not yet really, but folks over in knative/build land want to use it. started hacking on it a while ago: https://github.com/google/go-containerregistry/pull/261